Onion: multilayered concentric sphere and ellipsoid shapes

jeromefung opened the pull request:
### Description
Add shapes `onion` and `onion_ell` for multilayered concentric spheres and ellipsoids, respectively.

`onion` extends existing shapes `coated` and `coated2` to allow for arbitrary number of concentric layers, not just 2 or 3. (The upper limit is set by `MAX_NMAT`.) Similarly, `onion_ell` extends existing `ellipsoid` shape to allow for an arbitrary number of concentric ellipsoidal layers. These new shapes might be useful for modeling particles whose refractive index varies continuously by approximating the variation stepwise.

### Related issues
Fixes #293. (Confocal spheroids, briefly mentioned in #293, are not implemented.)

### Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist

- [x] I have read the [contributing guidelines](https://github.com/adda-team/adda/wiki/InstructionCommitters)
- [x] The new code complies with the existing [code style](https://github.com/adda-team/adda/wiki/CodeStyleGuide).
- [x] The code compiles correctly in all relevant regimes (at least, `make seq`)
- [x] The new code does not rely on any Fortran or C++ sources or is disabled by `NO_FORTRAN` or `NO_CPP` preprocessor macros, respectively.
- [x] The change neither adds or removes files; otherwise, these changes are reflected in `README.md` in corresponding folders.  
- [x] No warnings appear during debug compilation (at least, `make seq OPTIONS=DEBUG`, but better `devtools/build_debug`) or they are discussed below
- [x] Tests pass locally with my changes (at least, `sh comp2exec seq` in `tests/2exec`, but better `devtools/test_new [seq]`). If any errors appear, they are discussed below.
- [x] I have added tests that prove my fix is effective or that my feature works. And these tests pass. This includes new command line in suite files in `tests/2exec` (and potentially new ignore patterns). In some cases, it is desirable to add new tests to `tests/equiv`. 
- [x] I have added/extended necessary documentation (if appropriate). If suggesting changes to the manual, I have used "Track changes" in the doc file.
- [x] I have looked through all changes introduced by this pull request (line by line), using `git diff` or, better, some GUI tool, to ensure that no unexpected changes are introduced.

* Some tests in `comp2exec seq` fail, but same failures occur in `master`.
* Descriptions of the new shapes have been added to the manual. However, I did not attempt to update Fig. 3 (pictures of predefined shapes) since it is already very crowded and a rearrangement would be needed to fit in additional shapes.

myurkin committed:
Updates copyright string to year 2024. Fixes minor compiler warning. Updates script for Intel compiler with VS (without testing)

jeromefung committed:
Merge branch 'onion' of https://github.com/jeromefung/adda into onion

jeromefung committed:
onion works; checked for consistency with coated and coated2

jeromefung committed:
started onion_ell; done with param.c

jeromefung committed:
shape onion_ell implemented; passes comparison to ellipsoid when all layers have same index, and comparison to onion when yxratio=zxratio=1

jeromefung committed:
equiv test for onion

jeromefung committed:
forgot to update tests/equiv/README with new file

jeromefung committed:
updated documentation, ignores in comp2exec, suite includes onion and onion_ell

jeromefung committed:
restore reference binary path in comp2exec to windows binaries

jeromefung committed:
cleanup for style, whitespace, add ellipsoid/onion_ell comparison of CrossSec-X

jeromefung committed:
Merge branch 'adda-team:master' into onion

jeromefung committed:
Merge branch 'adda-team:master' into onion

jeromefung committed:
Fixed a few cases of previously missed spaces after commas in function arguments

myurkin commented on src/make_particle.c:
```diff
@@ -2211,6 +2271,38 @@ void MakeParticle(void)
+					// first test if inside the core, then test if in layers from the outside in
+					if (r2<=onion_r2[nlayers-2]) mat=nlayers-1; // inside core 
+					else {
+						for (layer_ctr=0;layer_ctr<(nlayers-1);layer_ctr++) {
+							if (r2>onion_r2[layer_ctr]) {
+								mat=layer_ctr;
+								break;
+							}
```
Have you considered any acceleration for such search? 

Actually, I tried running it with 250 layers, and the time for make_particle was always at least 10 times smaller that a single iteration. So definitely not a bottleneck. However, the code may even be neater if bsearch function is used.

jeromefung replied:
Respectfully, I'm not sure that using `bsearch` would significantly improve performance (given that `make_particle` is not an overall bottleneck in running ADDA) or would make this easier to read. Unless I'm missing something, `bsearch` would find an exact match between the `r2` value of a given dipole and an element of `onion_r2`, but that's not what is needed here. Moreover, using `bsearch` would require writing a comparison function.

myurkin reviewed (commented):
Jerome, as I said before, your contribution is really comprehensive. Thanks for that! I will have some minor changes here and there (mostly in help strings and such, I will commit them later), but here are couple of questions, which you may want to address.

jeromefung commented:
Thanks very much -- I'll respond to your comments over the next day or two.

myurkin committed:
Streamlines and improves the onion.sh and superellipsoid.sh:
- adds several new functions
- moves common parts to a new file common.sh
- comparison of shape files now ignores differences in shape names (if it is expected)
- adds processing of exit codes and possibility to stop on the first error
- increases refractive indices, adds -grid 16, but balances this with -eps 1 (keeping the computational time approximately the same)
- adds several tests to onion.sh with zero-thickness shells
- now both tests run clean

myurkin committed:
- updates shapes.svg to include two new shapes (instead deprecated coated2)
- adds hints on inserting SVG files into MS Word in svg_images.txt
- Changes to the manual:
* polishes the description of new shapes (including some changes in variable names)
* specifies coated2 as deprecated shape
* updates the default value of MAX_NMAT to 60
* updates affiliation of the first author
* adds specific note about an interim version of the manual

myurkin committed:
Improvements to tests/2exec:
- a bit cleaning in CMDIGNORE and OLDIGNORE
- makes test less stringent for "-int igt 3" (due to improvement of corresponding routines since 1.4.0)
- copies new test strings to other suites (haven't tested sparse, since it is all commented there)

myurkin committed:
- MAX_NMAT increases to 60 (to work out of the box for most applications)
- MAX_N_SH_PARMS is now calculated directly from MAX_NMAT, but extra check for its sufficiency is added to PARSE_FUNC(shape)
- Use of MAX_NMAT is now more informative in various help strings (`-h m` and for new shapes) due to use of new macro TO_STRING(A)
- adds deprecated comments and warnings for shape coated2
- increments ADDA version to 1.5.0-alpha2 (as was done in the manual)

myurkin commented:
I have added various improvement. Feel free to comment on them (or add anything). Apart of that, we only need to finalize `make_particle.c`

jeromefung committed:
streamline determination of number of layers in onion and onion_ell

jeromefung commented on src/make_particle.c:
```diff
@@ -1687,6 +1689,64 @@ void InitShape(void)
+				if (sh_pars[i]==0) {
+					nlayers=i+1;
+					break;
+					}
```
It was not my intention to allow any arguments to be 0 -- I was trying to determine the number of layers by counting the number of non-zro parameters in `sh_pars`. Looking again at this section of the code, I realized that there's a much simpler way: the variable `sh_Npars`. 

I've pushed a fix for this and will look at the bsearch issue by the end of the week.

jeromefung commented:
Your changes to files other than `make_particle.c` look good to me, and I think I've addressed your questions about `make_particle.c`. Please let me know if you have any other questions or concerns; I have no further issues that would preclude merging.

myurkin committed:
adds description of new file in test/equiv

myurkin committed:
A few updates to onion shapes:
- tests of input parameters is now inclusive, so equality of any bounds (including that to 0 and 1) is allowed
- cosmetic improvements to description strings
- onion_r2 is now dynamically allocated
- adds function DescendingSearch - the implementation is still trivial, but the overall code is simpler

myurkin commented:
Jerome, I have added a few (cosmetic) improvements. Can you please quickly check that the latest version works on your side? And then I will merge.
